RecoLocalCalo/HcalRecProducers/plugins/alpaka/Mahi.dev.cc

```diff
@@ -0,0 +1,1470 @@
+                //Loop over sample
+                for (auto sample : independent_group_elements_x(acc, nsamples)) {
+                  // conditions
+                  auto const id = gch < f01HEDigis.size()
+                                      ? f01HEDigis.ids()[gch]
+                                      : (gch < nchannelsf015 ? f5HBDigis.ids()[gch - f01HEDigis.size()]
+                                                             : f3HBDigis.ids()[gch - nchannelsf015]);
+                  auto const deltaT = gch >= f01HEDigis.size() && gch < nchannelsf015 ? timeSigmaHPD : timeSigmaSiPM;
+
+                  // compute hash for this did
+                  // hash needed to convert condition arrays order (based on Topo) into digi arrays order(based on FED)
+                  auto const did = DetId{id};
+                  auto const hashedId =
+                      did.subdetId() == HcalBarrel
+                          ? did2linearIndexHB(
+                                id, mahi.maxDepthHB(), mahi.firstHBRing(), mahi.lastHBRing(), mahi.nEtaHB())
+                          : did2linearIndexHE(id,
+                                              mahi.maxDepthHE(),
+                                              mahi.maxPhiHE(),
+                                              mahi.firstHERing(),
+                                              mahi.lastHERing(),
+                                              mahi.nEtaHE()) +
+                                mahi.offsetForHashes();
+
+                  // conditions based on the hash
+                  auto const recoPulseShapeId = recoParamView.ids()[hashedId];
+                  auto const* acc25nsVec = pulseShapeView.acc25nsVec()[recoPulseShapeId].data();
+                  auto const* diff25nsItvlVec = pulseShapeView.diff25nsItvlVec()[recoPulseShapeId].data();
+                  auto const* accVarLenIdxMinusOneVec =
+                      pulseShapeView.accVarLenIdxMinusOneVec()[recoPulseShapeId].data();
+                  auto const* diffVarItvlIdxMinusOneVec =
+                      pulseShapeView.diffVarItvlIdxMinusOneVec()[recoPulseShapeId].data();
+                  auto const* accVarLenIdxZeroVec = pulseShapeView.accVarLenIdxZEROVec()[recoPulseShapeId].data();
+                  auto const* diffVarItvlIdxZeroVec = pulseShapeView.diffVarItvlIdxZEROVec()[recoPulseShapeId].data();
+
+                  // offset output arrays
+                  auto* pulseMatrix = pulseMatrices + nsamples * npulses * gch;
+                  auto* pulseMatrixM = pulseMatricesM + nsamples * npulses * gch;
+                  auto* pulseMatrixP = pulseMatricesP + nsamples * npulses * gch;
+
+                  // amplitude per ipulse
+                  int const soi = soiSamples[gch];
+                  int const pulseOffset = pulseOffsets.offsets()[ipulse];
+                  auto const amplitude = amplitudes[gch * nsamples + pulseOffset + soi];
+
```

Comment:
If I read the code correctly, when `amplitude <= 0.f`, the final condition ignored all the computations and simply uses
```c++
                    pulseMatrix[ipulse * nsamples + sample] = 0.f;
                    pulseMatrixM[ipulse * nsamples + sample] = 0.f;
                    pulseMatrixP[ipulse * nsamples + sample] = 0.f;
```

Would it make sense to do here something like
```c++
                  if (amplitude <= 0.f) {
                    pulseMatrix[ipulse * nsamples + sample] = 0.f;
                    pulseMatrixM[ipulse * nsamples + sample] = 0.f;
                    pulseMatrixP[ipulse * nsamples + sample] = 0.f;
                    continue;
                  }
```
and skip the extra computations ?

Reply:
Done in 11aa37463e89e26d17c31a9e9d76e6f799b8a298